File: Data Structures/BinaryTrees/DiagonalPrint.java
import java.util.HashMap;
import java.util.Map.Entry;
import java.util.Vector;

public class DiagonalTraversal
{
    static class Node
    {
        int data;
        Node left;
        Node right;
        
        Node(int data)
        {
            this.data = data;
            left = null;
            right = null;
        }
    }

    static void diagonalPrintUtil(Node root, int d, HashMap<Integer, Vector<Integer>> diagnal_map)
    {
        if(root == null)
            return;
        Vector<Integer> k = diagonal_print.get(d);
        if (k == null)
        {
            k = new Vector<>();
            k.add(root.data);
        }
        else
        {
            k.add(root.data);
        }
        diagonal_map.put(d, k);
        diagonalPrintUtil(root.left, d+1, diagonal_map);
        diagonalPrintUtil(root.right, d, diagonal_map);
    }

    static void diagonalPrint(Node root)
    {
        HashMap<Integer, Vector<Integer>> diagonal_map = new HashMap<>();
        diagonalPrintUtil(root, 0, diagonal_map);
        System.out.println("diagonal - ");
        for(Entry<Integer, Vector<Integer>> entry : diagonalPrint.entrySet();)
        {
            System.out.println(entry.getValue());
        }
    }
}

File: Data Structures/BinaryTrees/DiagonalPrint.py
class Node:

    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def diagonalPrintUtil(root, d, diagonal_map):
    if root is None:
        return

    try:
        diagonal_map[d].append(root.data)
    except KeyError:
        diagonal_map[d] = [root.data]

    diagonalPrintUtil(root.left, d+1, diagonal_map)
    diagonalPrintUtil(root.right, d, diagonal_map)

def diagonalPrint(root):

    diagonal_map = dict()
    diagonalPrintUtil(root, 0, diagonal_map)
    print("Diagonal traversal is : ")
    for i in diagonal_map:
        for j in diagonal_map[i]:
            print(j)
        print " "

root = Node(8)

diagonalPrint(root)
